test/circuit_string.test.ts

```diff
@@ -0,0 +1,164 @@
+        expect(str.contains(contained_str)).toEqual(new Bool(true))
+
+        Circuit.runAndCheck(() => {
+          const str = CircuitString8.fromString('abcd')
+          const contained_str = CircuitString.fromString('ab')
+          expect(str.contains(contained_str)).toEqual(new Bool(true))
+        });
+      });
+
+      test('string may contain string8', () => {
+        const str = CircuitString.fromString('abcd')
+        const contained_str = CircuitString8.fromString('ab')
+        expect(str.contains(contained_str)).toEqual(new Bool(true))
+
+        Circuit.runAndCheck(() => {
+          const str = CircuitString.fromString('abcd')
+          const contained_str = CircuitString8.fromString('ab')
+          expect(str.contains(contained_str)).toEqual(new Bool(true))
+        });
+      });
+    });
+  });
+
+  describe('#toString', () => {
+    test('serializes to string', () => {
+      const js_str = 'Everything we hear is an opinion, not a fact. Everything we see is a perspective, not the truth';
+      const str = CircuitString.fromString(js_str)
+      expect(str.toString()).toBe(js_str);
+
+      Circuit.runAndCheck(() => {
+        const js_str = 'Everything we hear is an opinion, not a fact. Everything we see is a perspective, not the truth';
+        const str = CircuitString.fromString(js_str)
+        expect(str.toString()).toBe(js_str);
+      });
+    });
+  });
+
+  describe('#substring', () => {
+    test('selects substring', () => {
+      const str = CircuitString.fromString('Everything we hear is an opinion, not a fact. Everything we see is a perspective, not the truth')
+      expect(str.substring(46, 80).toString()).toBe('Everything we see is a perspective');
+
+      Circuit.runAndCheck(() => {
+        const str = CircuitString.fromString('Everything we hear is an opinion, not a fact. Everything we see is a perspective, not the truth')
+        expect(str.substring(46, 80).toString()).toBe('Everything we see is a perspective');
+      });
+    });
+  });
+
+  describe('#append', () => {
+    test('appends 2 strings', () => {
+      const str1 = CircuitString.fromString('abcd')
+      const str2 = CircuitString.fromString('efgh')
+      expect(str1.append(str2).toString()).toBe('abcdefgh');
+
+      Circuit.runAndCheck(() => {
+        const str1 = CircuitString.fromString('abcd')
+        const str2 = CircuitString.fromString('efgh')
+        expect(str1.append(str2).toString()).toBe('abcdefgh');
+      });
+    });
+  });
+
+  describe('CircuitString8', () => {
+    // This test passes but it should not
+    test('prop size is respected', () => {
+      Circuit.runAndCheck(() => {
+        const str = CircuitString8.fromString('More than eight chars')
+        expect(str.toString()).toBe('More than eight chars')
+      });
+    })
+  });
+
+  describe('with invalid input', () => {
+    // This test passes but it should not
+    test('cannot use a character out of range', () => {
```

Comment:
@mitschabaude Similar issue here, I would expect `Field(10000)` to be an invalid Character and for `#check` to throw an error on init for a character.